### ISSUE TITLE
string-lit-as-bytes fix producing incorrect fix

### ISSUE THREAD

**Issue by u9g** [rust-lang/rust-clippy#11729]
### Summary

Lint fix proposes code which fails to compile.

### Reproducer

I tried this code:

```rust
#![warn(clippy::string_lit_as_bytes)]

fn reader(_: impl std::io::Read) {}

fn main() {
    reader("Hello".as_bytes());
}
```

I expected to see this happen: To get a compile-able lint fix.

Instead, this happened: The fix suggested is an error.

### Version

```text
rustc 1.73.0 (cc66ad468 2023-10-03)
binary: rustc
commit-hash: cc66ad468955717ab92600c770da8c1601a4ff33
commit-date: 2023-10-03
host: x86_64-unknown-linux-gnu
release: 1.73.0
LLVM version: 17.0.2
```

### Additional Labels

@rustbot label +I-suggestion-causes-error
